pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java

```diff
@@ -277,4 +253,112 @@ public DataTable processQuery(ServerQueryRequest queryRequest, ExecutorService e
+    }
+
+    TimerContext.Timer segmentPruneTimer = timerContext.startNewPhaseTimer(ServerQueryPhase.SEGMENT_PRUNING);
+    List<IndexSegment> selectedSegments = _segmentPrunerService.prune(indexSegments, queryContext);
+    segmentPruneTimer.stopAndRecord();
+    int numSelectedSegments = selectedSegments.size();
+    LOGGER.debug("Matched {} segments after pruning", numSelectedSegments);
+    if (numSelectedSegments == 0) {
+      // Only return metadata for streaming query
+      DataTable dataTable = enableStreaming ? new DataTableImplV2() : DataTableUtils.buildEmptyDataTable(queryContext);
+      Map<String, String> metadata = dataTable.getMetadata();
+      metadata.put(DataTable.TOTAL_DOCS_METADATA_KEY, String.valueOf(numTotalDocs));
+      metadata.put(DataTable.NUM_DOCS_SCANNED_METADATA_KEY, "0");
+      metadata.put(DataTable.NUM_ENTRIES_SCANNED_IN_FILTER_METADATA_KEY, "0");
+      metadata.put(DataTable.NUM_ENTRIES_SCANNED_POST_FILTER_METADATA_KEY, "0");
+      metadata.put(DataTable.NUM_SEGMENTS_PROCESSED, "0");
+      metadata.put(DataTable.NUM_SEGMENTS_MATCHED, "0");
+      return dataTable;
+    } else {
+      TimerContext.Timer planBuildTimer = timerContext.startNewPhaseTimer(ServerQueryPhase.BUILD_QUERY_PLAN);
+      Plan queryPlan = enableStreaming ? _planMaker
+          .makeStreamingInstancePlan(selectedSegments, queryContext, executorService, responseObserver, endTimeMs)
+          : _planMaker.makeInstancePlan(selectedSegments, queryContext, executorService, endTimeMs);
+      planBuildTimer.stopAndRecord();
+
+      TimerContext.Timer planExecTimer = timerContext.startNewPhaseTimer(ServerQueryPhase.QUERY_PLAN_EXECUTION);
+      DataTable dataTable = queryPlan.execute();
+      planExecTimer.stopAndRecord();
+
+      // Update the total docs in the metadata based on the un-pruned segments
+      dataTable.getMetadata().put(DataTable.TOTAL_DOCS_METADATA_KEY, Long.toString(numTotalDocs));
+
+      return dataTable;
+    }
+  }
+
+  /**
+   * Handles the subquery in the given query.
+   * <p>Currently only supports subquery within IN_PARTITIONED_SUBQUERY function.
+   */
+  private void handleSubquery(QueryContext queryContext, List<IndexSegment> indexSegments, TimerContext timerContext,
+      ExecutorService executorService, long endTimeMs)
+      throws Exception {
+    FilterContext filter = queryContext.getFilter();
+    if (filter != null) {
+      handleSubquery(filter, indexSegments, timerContext, executorService, endTimeMs);
+    }
+  }
+
+  /**
+   * Handles the subquery in the filter.
+   * <p>Currently only supports subquery within IN_PARTITIONED_SUBQUERY function.
+   */
+  private void handleSubquery(FilterContext filter, List<IndexSegment> indexSegments, TimerContext timerContext,
+      ExecutorService executorService, long endTimeMs)
+      throws Exception {
+    Predicate predicate = filter.getPredicate();
+    if (predicate != null) {
```

Comment:
docs, isn't it better to check for filter.getChildren and invert the if statement?

Reply:
Done